File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/Translate.java
package net.gnehzr.tnoodle.server.webscrambles;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Map;
import java.util.Locale;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.Set;

import net.gnehzr.tnoodle.utils.EnvGetter;
import net.gnehzr.tnoodle.utils.Utils;

import org.yaml.snakeyaml.constructor.Constructor;
import org.yaml.snakeyaml.Yaml;

import static net.gnehzr.tnoodle.utils.GwtSafeUtils.azzert;

public class Translate {
    private Translate() {}

    private static final Logger l = Logger.getLogger(Translate.class.getName());
    private static final Locale BASE_LOCALE = Locale.forLanguageTag("en");
    public static final Locale DEFAULT_LOCALE = Locale.forLanguageTag(EnvGetter.getenv("TNOODLE_DEFAULT_LOCALE", BASE_LOCALE.toLanguageTag()));

    private static final HashMap<Locale, HashMap<String, ?>> TRANSLATIONS = new HashMap<Locale, HashMap<String, ?>>();
    static {
        HashMap<String, HashMap<String, ?>> translationsByLanguageTag = new HashMap<String, HashMap<String, ?>>();
        File i18nDir = new File(Utils.getResourceDirectory(), "i18n");
        File[] ymlFiles = i18nDir.listFiles();
        for(File ymlFile : ymlFiles) {
            HashMap<String, ?> data = null;
            try {
                Yaml yaml = new Yaml(new Constructor(HashMap.class));
                deepMerge(translationsByLanguageTag, yaml.load(new FileInputStream(ymlFile)));
            } catch(FileNotFoundException e) {
                l.log(Level.SEVERE, "Error loading translation", e);
                azzert(false);
            }
        }

        for(String languageTag : translationsByLanguageTag.keySet()) {
            Locale locale = Locale.forLanguageTag(languageTag);

            TRANSLATIONS.put(locale, translationsByLanguageTag.get(languageTag));
        }
    }

    public static Set<Locale> getLocales() {
        return TRANSLATIONS.keySet();
    }

    private static HashMap<String, ?> getTranslationsFor(Locale locale) {
        HashMap<String, ?> translations = TRANSLATIONS.get(locale);
        azzert(translations != null, "No locale matching given locale: " + locale.toLanguageTag() + " found");
        return translations;
    }

    @SuppressWarnings("unchecked")
    private static String translate(String key, Locale locale, boolean azzertIfInvalid) {
        Object translation = getTranslationsFor(locale);

        String[] parts = key.split("\\.");
        for(String part : parts) {
            if(!(translation instanceof HashMap)) {
                if(azzertIfInvalid) {
                    azzert(false, locale.toLanguageTag() + " couldn't find part " + part + " of " + key);
                } else {
                    return null;
                }
            }

            translation = ((HashMap<String, ?>) translation).get(part);
        }

        if(translation == null) {
            if(azzertIfInvalid) {
                azzert(false, locale.toLanguageTag() + " translation key " + key + " not found");
            } else {
                return null;
            }
        }

        if(!(translation instanceof String)) {
            if(azzertIfInvalid) {
                azzert(false, locale.toLanguageTag() + " translation key " + key + " is of type " + translation.getClass() + ", but we were expecting String.");
            } else {
                return null;
            }
        }

        return (String) translation;
    }

    public static String translate(String key, Locale locale) {
        return translate(key, locale, new HashMap<String, String>());
    }

    public static String translate(String key, Locale locale, Map<String, String> substitutions) {
        // Attempt to translate in the given locale.
        String translation = translate(key, locale, false);
        if(translation == null) {
            // If we couldn't find a translation in the given locale, fall back to the base locale.
            translation = translate(key, BASE_LOCALE, true);
        }

        translation = interpolate(translation, substitutions);
        return translation;
    }

    // Interpolate translation keys in the same way Ruby on Rails does.
    // See: http://guides.rubyonrails.org/i18n.html#passing-variables-to-translations
    private static String interpolate(String interpolateMe, Map<String, String> substitutions) {
        // Copy the given Map, as we're going to mutate it later on.
        substitutions = new HashMap<String, String>(substitutions);

        // Now interpolate variables in the input string.
        //
        // Find anything that looks like %{...}, unless the percent sign is escaped as in %%{...}
        Pattern templatePattern = Pattern.compile("(?<!%)%\\{([^}]+)\\}");
        Matcher matcher = templatePattern.matcher(interpolateMe);
        StringBuffer sb = new StringBuffer();
        while(matcher.find()) {
            String key = matcher.group(1);
            String replacement = substitutions.remove(key);
            azzert(replacement != null, String.format("Substitution for key: %s not found", key));
            matcher.appendReplacement(sb, replacement);
        }
        matcher.appendTail(sb);
        azzert(substitutions.size() == 0, String.format("Unused substitution values: %s", substitutions.keySet()));
        return sb.toString();
    }

    // Copied from https://gist.github.com/aslakhellesoy/3858814
    @SuppressWarnings({"unchecked", "rawtypes"})
    private static final Map deepMerge(Map original, Map newMap) {
        for(Object key : newMap.keySet()) {
            if(newMap.get(key) instanceof Map && original.get(key) instanceof Map) {
                Map originalChild = (Map) original.get(key);
                Map newChild = (Map) newMap.get(key);
                original.put(key, deepMerge(originalChild, newChild));
            } else {
                original.put(key, newMap.get(key));
            }
        }
        return original;
    }
}

File: webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/Translate.kt
package org.worldcubeassociation.tnoodle.server.webscrambles

import org.worldcubeassociation.tnoodle.server.webscrambles.exceptions.TranslationException
import org.yaml.snakeyaml.LoaderOptions
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.Constructor

import java.util.Locale

object Translate {
    private val BASE_LOCALE = Locale.forLanguageTag("en")
    val DEFAULT_LOCALE = System.getenv("TNOODLE_DEFAULT_LOCALE")
        ?.let { Locale.forLanguageTag(it) } ?: BASE_LOCALE

    val TRANSLATIONS = loadTranslationResources()

    val TRANSLATED_LOCALES
        get() = TRANSLATIONS.keys

    val LOCALES_BY_LANG_TAG = TRANSLATED_LOCALES.associateBy { it.toLanguageTag() }

    private fun loadTranslationResources(): Map<Locale, Map<String, *>> {
        val yaml = Yaml(Constructor(HashMap::class.java, LoaderOptions()))
        val locales = Locale.getAvailableLocales() + DEFAULT_LOCALE

        val loadedMaps = locales
            .mapNotNull { javaClass.getResourceAsStream("/i18n/${it.toLanguageTag()}.yml") }
            .map { yaml.load<Map<String, Map<String, *>>>(it) }

        val mergedMaps = loadedMaps
            .takeIf { it.isNotEmpty() }
            ?.reduce { a, b -> a + b }
            ?.mapKeys { Locale.forLanguageTag(it.key) }

        return mergedMaps ?: emptyMap()
    }

    private fun getTranslationsFor(locale: Locale): Map<String, *> {
        return TRANSLATIONS[locale]
            ?: TranslationException.error("No locale matching given locale: ${locale.toLanguageTag()} found")
    }

    private fun translateNullable(key: String, locale: Locale): String? {
        val baseTranslations = getTranslationsFor(locale)
        val parts = key.split(".")

        return descendKeys(baseTranslations, parts)
    }

    private fun translateWithError(key: String, locale: Locale): String {
        val baseTranslations = getTranslationsFor(locale)
        val parts = key.split(".")

        val translation = descendKeys(baseTranslations, parts) {
            "${locale.toLanguageTag()} couldn't find part $it of $key"
        } ?: TranslationException.error("${locale.toLanguageTag()} translation key $key not found")

        return translation as? String
            ?: TranslationException.error("${locale.toLanguageTag()} translation key $key is of type ${translation.javaClass}, but we were expecting String.")
    }

    private tailrec fun descendKeys(
        translationGroup: Map<*, *>,
        parts: List<String>,
        errorProvider: ((String) -> String)? = null
    ): String? {
        if (parts.isEmpty()) {
            return null
        }

        val nextPart = parts.first()
        val remainingParts = parts.drop(1)

        val nextGroup = translationGroup[nextPart]

        if (remainingParts.isEmpty()) {
            return nextGroup as? String
        }

        if (nextGroup !is Map<*, *>) {
            return errorProvider?.invoke(nextPart)
                ?.let { TranslationException.error(it) }
        }

        return descendKeys(nextGroup, remainingParts)
    }

    fun translate(key: String, locale: Locale, substitutions: Map<String, String> = mapOf()): String {
        // Attempt to translate in the given locale.
        val translation = translateNullable(key, locale)
            // If we couldn't find a translation in the given locale, fall back to the base locale.
            ?: translateWithError(key, BASE_LOCALE)

        return interpolate(translation, substitutions)
    }

    operator fun invoke(key: String, locale: Locale, substitutions: Map<String, String> = mapOf()) =
        translate(key, locale, substitutions)

    // Interpolate translation keys in the same way Ruby on Rails does.
    // See: http://guides.rubyonrails.org/i18n.html#passing-variables-to-translations
    private fun interpolate(interpolateMe: String, substitutions: Map<String, String>): String {
        // Copy the given Map, as we're going to mutate it later on.
        val mutateSubstitutions = substitutions.toMutableMap()

        // Now interpolate variables in the input string.
        //
        // Find anything that looks like %{...}, unless the percent sign is escaped as in %%{...}
        val pattern = Regex("(?<!%)%\\{([^}]+)}")
        val interpolated = pattern.replace(interpolateMe) {
            val key = it.groups[1]?.value
                ?: TranslationException.error("Translation Regex broken: Group without content for $interpolateMe")

            mutateSubstitutions.remove(key)
                ?: TranslationException.error("Substitution for key: $key not found")
        }

        assert(mutateSubstitutions.isEmpty()) { "Unused substitution values: ${mutateSubstitutions.keys}" }

        return interpolated
    }
}
